cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -525,12 +591,17 @@ class self_comparator {
    * `F(i,j)` returns true if and only if row `i` compares lexicographically less than row `j`.
    *
    * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
+   * @tparam PhysicalElementComparator A relational comparator functor, defaults to `NaN` aware
+   * relational comparator that evaluates `NaN` as greater than standard values.
    */
-  template <typename Nullate>
-  less_comparator<device_row_comparator<Nullate>> device_comparator(Nullate nullate = {}) const
+  template <typename Nullate,
+            typename PhysicalElementComparator = sorting_physical_element_comparator>
+  less_comparator<device_row_comparator<Nullate, PhysicalElementComparator>> device_comparator(
+    Nullate nullate = {}) const
   {
-    return less_comparator<device_row_comparator<Nullate>>{device_row_comparator<Nullate>(
-      nullate, *d_t, *d_t, d_t->depths(), d_t->column_order(), d_t->null_precedence())};
+    return less_comparator<device_row_comparator<Nullate, PhysicalElementComparator>>{
+      device_row_comparator<Nullate, PhysicalElementComparator>(
+        nullate, *d_t, *d_t, d_t->depths(), d_t->column_order(), d_t->null_precedence())};
```

Comment:
CTAD?

Reply:
Maybe I'm confused what you're asking about, but with the default arguments I wasn't able to get it to work when letting the compiler deduce the template arguments.

Reply:
I remember CTAD failing here when building strongly-typed comparators. I don't recall exactly what went wrong.

Reply:
To complicate this a bit more, in overlaying some of these changes onto #9452 I realized less_comparator is hard coded. While the use case is very specific, we'll still need a way to get either the `strong_index_comparator_adapter` back or template the `less_comparator`.

Here is my proposed solutions, which would you prefer:
Extending the templating of the existing function
```
  template <typename Nullate,
            typename PhysicalElementComparator = sorting_physical_element_comparator,
            typename ComparatorInterpreter     = less_comparator<strong_index_comparator_adapter<
              device_row_comparator<Nullate, PhysicalElementComparator>>>>
  ComparatorInterpreter device_comparator(Nullate nullate = {}) const noexcept
  {
    return ComparatorInterpreter{
      device_row_comparator<Nullate, PhysicalElementComparator>(nullate,
                                                                *d_left_table,
                                                                *d_right_table,
                                                                d_left_table->depths(),
                                                                d_left_table->column_order(),
                                                                d_left_table->null_precedence())};
  }
```

Extra function, allow caller to construct the interpreter from the `strong_index_comparator_adapter`
```
  template <typename Nullate,
            typename PhysicalElementComparator = sorting_physical_element_comparator>
  auto device_adapter(Nullate nullate = {}) const noexcept
  {
    return strong_index_comparator_adapter{
      device_row_comparator<Nullate, PhysicalElementComparator>(nullate,
                                                                *d_left_table,
                                                                *d_right_table,
                                                                d_left_table->depths(),
                                                                d_left_table->column_order(),
                                                                d_left_table->null_precedence())};
  }
```

I wanted to get a second opinion, my gut says the extra function option to avoid unreadable template stacks like this one from my test code
```
    auto device_comparator = table_comparator.device_comparator<
      nullate::DYNAMIC,
      cudf::experimental::row::lexicographic::physical_element_comparator,
      cudf::experimental::row::lexicographic::less_equivalent_comparator<
        cudf::experimental::row::lexicographic::strong_index_comparator_adapter<
          cudf::experimental::row::lexicographic::device_row_comparator<
            nullate::DYNAMIC,
            cudf::experimental::row::lexicographic::physical_element_comparator>>>>(
      nullate::DYNAMIC{out.has_nulls()});
```

Reply:
I'm not following. This is sounding far more complex than it should be. 

> While the use case is very specific, we'll still need a way to get either the strong_index_comparator_adapter back or template the less_comparator.

Can you explain further? 

I _think_ if I'm understanding you correctly, you need a way to get a `less_comparator` or `less_equal_comparator`. The solution is to rename `self_comparator::device_comparator` to `self_comparator::less` and add `self_comparator::less_equal`.

Reply:
The reason CTAD didn't work is because `less_comparator` is an alias and CTAD for template aliases wasn't added until C++20. 

That's easy enough to work around by instead making it inherit from `weak_ordering_comparator_impl`, but is more annoying than it needs to be because CTAD doesn't seem to work with inherited ctors. https://godbolt.org/z/E1fovs1oc

So here's a thing that works: https://godbolt.org/z/9Eo9YvsTz

Reply:
Thanks again for the example Jake. Added the explicit constructor and CTAD cleans it up considerably.